### DISCUSSION TITLE
How to add a finalizer for an effect action?

### DISCUSSION THREAD

**Discussion by MuhammedZakir** [haskell-effectful/effectful#85] [Q&A] [answered]
I have two effects:

```haskell
data E1 :: Effect where
  GetStuff :: E1 m Stuff
  DiscardStuff :: Stuff -> E1 m ()

-- 'E1' effect is necessary for 'E2'
data E2 :: Effect where
  FuncThatUseStuff :: E2 m ()
```

There are two ways `E1` can be used:
1. user discards the `Stuff` they got:
```haskell
use1 = do
  s <- getStuff
  ...
  discardStuff s
```

2. if user doesn't discard any `Stuff` they got, then I will have to handle it:
```haskell
use2 = do
  s1 <- getStuff
  s2 <- getStuff
  ...
  discardStuff s1
  -- user didn't discard `s2`, so I'll have to handle it
```

The problem is, I don't know how attach a finalizer without using continuation / passing function:
```
handler stuffs = reinterpret (runState stuffs) $ \_ -> \case
  GetStuff  -> do
    s <- pop
    save s
    pure s
    -- HERE: after `s` is used, I want to handle it if user doesn't call `discardStuff`.

  DiscardStuff s -> do
    when (inSavedStuff s) $ delete s
```

---

I am using free monad interpreter atm. In it, I did something like:
```haskell
run = \case
  ...
  GetStuff nextFn -> do
    s <- pop
    save s
    res <- run (nextFn s)
      `finally` (handleIfNotDiscarded s)
    deleteSaved s
    pure res
  ...
```

**Comment by mmhat** [comment, marked as answer]
It seems like [resourcet-effectful](https://hackage.haskell.org/package/resourcet-effectful) would work for. Just [reinterpret](https://hackage.haskell.org/package/effectful-core-2.0.0.0/docs/Effectful-Dispatch-Dynamic.html#v:reinterpret) `E1` in terms of the `Resource` effect.

**Reply by MuhammedZakir** [reply]
Thanks!
